File: cane-cpp/cane-core/src/constraints/path.cpp
#include "solvers.hpp"

namespace cane::constraints {
namespace {
void include(std::optional<float>& residual, float value) {
    if (!std::isfinite(value)) throw Error(ErrorCode::non_finite, "apply", "Path diagnostic residual is non-finite.");
    residual = residual ? std::max(*residual, value) : value;
}
}
void solve_path(Context& context, std::size_t index) {
    using P = animation::ConstraintProperty;
    auto& pose = context.pose; auto& scratch = context.scratch;
    pose.diagnostics.at(index) = std::monostate{};
    const auto& d = std::get<detail::PathDefinition>(pose.model().constraints.at(index).definition);
    const auto& parameters = pose.constraints[index];
    const auto target_bone = pose.model().slots[d.target_slot].bone;
    if (!pose.active_constraints[index] || !pose.active_bones[target_bone]) return;
    const auto selected = pose.slots[d.target_slot].attachment;
    if (!selected || pose.model().attachments[*selected].kind != detail::AttachmentKind::path) return;
    const auto mr = std::clamp(parameters[P::mix_rotate], 0.0f, 1.0f);
    const auto mx = std::clamp(parameters[P::mix_x], 0.0f, 1.0f), my = std::clamp(parameters[P::mix_y], 0.0f, 1.0f);
    if (mr == 0 && mx == 0 && my == 0) return;
    const auto& attachment = pose.model().attachments[*selected];
    pose.source_vertices(*selected, scratch.path_vertices);
    auto& path = scratch.path; path.reset(scratch.path_vertices, attachment.closed, attachment.constant_speed, attachment.lengths);
    if (!path.available()) return;

    // No pose write occurs during this pass: every current matrix, including
    // ancestors and descendants in the same authored list, sees one snapshot.
    const auto& bones = pose.data().bones(); const auto count = d.bones.size();
    const auto tangent = d.rotate_mode == detail::RotateMode::tangent, chain_scale = d.rotate_mode == detail::RotateMode::chain_scale;
    const auto spacing_count = tangent && count != 0 ? count - 1 : count;
    const auto spaces_count = tangent ? count : count + 1; float total = 0;
    for (std::size_t i = 0; i < count; ++i) {
        const auto bone = d.bones[i]; scratch.path_lengths[bone] = std::max(bones[bone].length, 0.0f) * x_scale(pose.world[bone]);
        if (i < spacing_count) total += scratch.path_lengths[bone];
    }
    const auto spacing = parameters[P::spacing];
    auto distance = parameters[P::position] * (d.position_mode == detail::PositionMode::percent ? path.length() : 1);
    const auto offset = parameters.path_rotation != 0 && pose.world[target_bone].determinant() < 0 ? -parameters.path_rotation : parameters.path_rotation;
    auto& staged = scratch.path_desired; auto& driven = scratch.path_driven; std::fill(driven.begin(), driven.end(), 0);
    PathDiagnostic diagnostic; diagnostic.mix_rotate = mr; diagnostic.mix_x = mx; diagnostic.mix_y = my;
    std::optional<PathSample> previous;
    for (const auto bone : d.bones) {
        const auto setup_length = std::max(bones[bone].length, 0.0f), world_length = scratch.path_lengths[bone]; float increment = 0;
        switch (d.spacing_mode) {
            case detail::SpacingMode::length: increment = setup_length > 0 ? std::max(setup_length + spacing, 0.0f) * world_length / setup_length : spacing; break;
            case detail::SpacingMode::fixed: increment = setup_length > 0 ? spacing * world_length / setup_length : spacing; break;
            case detail::SpacingMode::percent: increment = spacing * path.length(); break;
            case detail::SpacingMode::proportional: increment = total > 0 ? (setup_length > 0 ? world_length : spacing) / total * spacing * path.length() : spacing * path.length() / static_cast<float>(std::max(spaces_count, std::size_t{1})); break;
        }
        const auto sample = previous ? previous : path.at(distance), next = path.at(distance + increment);
        previous.reset(); distance += increment;
        if (!sample || !next) continue;
        const auto point = sample->position, ahead = next->position; const auto current = pose.world[bone]; const auto old_angle = x_angle(current);
        const auto chain_angle = static_cast<float>(std::atan2(static_cast<double>(ahead.y - point.y), ahead.x - point.x)) * radians_to_degrees;
        const auto desired_angle = (tangent ? sample->tangent : chain_angle) + offset;
        const auto angle = old_angle + wrap_degrees(desired_angle - old_angle) * mr;
        auto scale_x = x_scale(current); const auto scale_y = y_scale(current);
        const auto target_length = cane::hypot(ahead.x - point.x, ahead.y - point.y);
        if (chain_scale && bones[bone].length != 0) scale_x += (target_length / std::abs(bones[bone].length) - scale_x) * mr;
        const auto shear_y = y_angle(current) - old_angle - 90;
        const auto desired = Affine::from_local({current.tx + (point.x - current.tx) * mx, current.ty + (point.y - current.ty) * my, angle, scale_x, scale_y, 0, shear_y});
        if (!desired.finite()) throw Error(ErrorCode::non_finite, "apply", "Path world matrix is non-finite.", "bones", bones[bone].id);
        staged[bone] = desired; driven[bone] = 1; ++diagnostic.driven;
        if (mx != 0 || my != 0) include(diagnostic.translation_residual, cane::hypot(mx == 0 ? 0 : desired.tx - point.x, my == 0 ? 0 : desired.ty - point.y));
        if (mr != 0) include(diagnostic.rotation_residual, std::abs(wrap_degrees(x_angle(desired) - desired_angle)));
        if (chain_scale && mr != 0 && bones[bone].length != 0) include(diagnostic.scale_residual, std::abs(x_scale(desired) - target_length / std::abs(bones[bone].length)));
        if (!tangent && !chain_scale && parameters.path_rotation == 0 && mr != 0) {
            const auto rotated = rotate_axes(current, wrap_degrees(chain_angle - old_angle));
            const Point tip{point.x + bones[bone].length * rotated.a, point.y + bones[bone].length * rotated.b};
            previous = PathSample{{ahead.x + (tip.x - ahead.x) * mr, ahead.y + (tip.y - ahead.y) * mr}, next->tangent};
        }
    }
    // Skeleton-order commit preserves every stored driven world, then recomposes
    // only affected non-driven descendants. Unrelated prior solver results survive.
    auto& affected = scratch.affected; std::fill(affected.begin(), affected.end(), 0);
    for (std::size_t bone = 0; bone < bones.size(); ++bone) {
        if (driven[bone]) { affected[bone] = 1; context.set_world(bone, staged[bone]); }
        else if (bones[bone].parent_index && affected[*bones[bone].parent_index]) {
            affected[bone] = 1; pose.world[bone] = context.compose(bone, pose.locals[bone]);
            if (!pose.world[bone].finite()) throw Error(ErrorCode::non_finite, "apply", "Path hierarchy is non-finite.", "bones", bones[bone].id);
        }
    }
    pose.diagnostics[index] = diagnostic;
}
}

File: cane-csharp/cane-core/Runtime/Constraints/Path.cs
using System;
using System.Collections.Generic;
using Cane.Animation;
using Cane.Format;
using Cane.Geometry;

namespace Cane.Constraints
{
    internal sealed partial class PathSampler
    {
        private readonly Curve[] curves;
        private readonly float[] lengths;
        private readonly float[][]? detailed;
        private readonly bool closed;
        internal bool Closed => closed;
        internal float Length => lengths.Length == 0 ? 0 : lengths[lengths.Length - 1];
        internal PathSampler(float[] world, Json path)
        {
            closed = path.B("closed"); int knots = world.Length / 6, count = closed ? knots : knots - 1;
            curves = new Curve[count]; lengths = new float[count]; bool constant = path.B("constantSpeed");
            if (constant) detailed = new float[count][];
            Point Read(int knot, int point) { int i = knot * 6 + point * 2; return new Point(world[i], world[i + 1]); }
            bool authored = !constant && path["lengths"].ArrayOrEmpty.Count == knots; float previous = 0;
            foreach (Json value in path["lengths"].ArrayOrEmpty) { float length = value.Float; if (!Numeric.Finite(length) || length < previous) authored = false; previous = length; }
            float sum = 0;
            for (int i = 0; i < count; i++)
            {
                int next = (i + 1) % knots; var c = new Curve(Read(i, 1), Read(i, 2), Read(next, 0), Read(next, 1)); curves[i] = c;
                if (authored) lengths[i] = path["lengths"][i].Float;
                else { float[] sampled = c.Lengths(constant ? 4 : 24); sum += sampled[sampled.Length - 1]; lengths[i] = sum; }
                if (detailed != null) detailed[i] = c.Lengths(10);
            }
        }
        internal Sample? At(float distance)
        {
            if (!Numeric.Finite(distance) || !Numeric.Finite(Length) || Length <= 0) return null;
            if (closed) { distance %= Length; if (distance < 0) distance += Length; }
            else if (distance < 0 || distance > Length)
            {
                bool start = distance < 0; Curve c = start ? curves[0] : curves[curves.Length - 1]; Point end = start ? c.P0 : c.P3, from = start ? c.P1 : c.P2;
                float dx = end.X - from.X, dy = end.Y - from.Y, length = Numeric.Hypot(dx, dy); if (length == 0) return null;
                float amount = start ? -distance : distance - Length; dx /= length; dy /= length;
                return new Sample(new Point(end.X + dx * amount, end.Y + dy * amount), Numeric.Atan2(dy, dx) * Numeric.RadToDeg);
            }
            int index = 0; while (index + 1 < lengths.Length && lengths[index] < distance) index++;
            float before = index == 0 ? 0 : lengths[index - 1], span = lengths[index] - before;
            float t = span == 0 ? 0 : Numeric.Clamp((distance - before) / span, 0, 1);
            if (detailed != null)
            {
                float[] lookup = detailed[index]; float target = t * lookup[lookup.Length - 1]; int segment = 0;
                while (segment + 1 < lookup.Length && lookup[segment] < target) segment++;
                float a = segment == 0 ? 0 : lookup[segment - 1], length = lookup[segment] - a;
                float ratio = length == 0 ? 0 : Numeric.Clamp((target - a) / length, 0, 1); t = (segment + ratio) / 10;
            }
            return new Sample(curves[index].Point(t), curves[index].Angle(t));
        }
        internal readonly struct Sample
        {
            internal readonly Point Position;
            internal readonly float Tangent;
            internal Sample(Point position, float tangent) { Position = position; Tangent = tangent; }
        }
        private readonly struct Curve
        {
            internal readonly Point P0, P1, P2, P3;
            internal Curve(Point p0, Point p1, Point p2, Point p3) { P0 = p0; P1 = p1; P2 = p2; P3 = p3; }
            internal Point Point(float t)
            {
                float u = 1 - t, a = u * u * u, b = 3 * u * u * t, c = 3 * u * t * t, d = t * t * t;
                return new Point(a * P0.X + b * P1.X + c * P2.X + d * P3.X, a * P0.Y + b * P1.Y + c * P2.Y + d * P3.Y);
            }
            internal float Angle(float t)
            {
                float u = 1 - t, dx = 3 * u * u * (P1.X - P0.X) + 6 * u * t * (P2.X - P1.X) + 3 * t * t * (P3.X - P2.X);
                float dy = 3 * u * u * (P1.Y - P0.Y) + 6 * u * t * (P2.Y - P1.Y) + 3 * t * t * (P3.Y - P2.Y);
                return (dx != 0 || dy != 0 ? Numeric.Atan2(dy, dx) : Numeric.Atan2(P3.Y - P0.Y, P3.X - P0.X)) * Numeric.RadToDeg;
            }
            internal float[] Lengths(int segments)
            {
                var values = new float[segments]; Point previous = P0; float total = 0;
                for (int i = 0; i < segments; i++) { Point p = Point((float)(i + 1) / segments); total += Numeric.Hypot(p.X - previous.X, p.Y - previous.Y); values[i] = total; previous = p; }
                return values;
            }
        }
    }
    internal static class PathConstraint
    {
        internal static void Solve(Pose pose, Json c, Affine root)
        {
            int slot = pose.Data.SlotIndex[c.S("targetSlotId")], targetBone = pose.Data.SlotData[slot].Bone;
            if (!pose.IsActive(pose.Data.BoneData[targetBone].Id, "boneIds")) return;
            int attachment = pose.Slots[slot].ResolvedAttachment; if (attachment < 0 || pose.Data.AttachmentData[attachment].Kind != "path") return;
            AttachmentData a = pose.Data.AttachmentData[attachment]; var path = new PathSampler(Rendering.WorldVertices(pose, a), a.Raw); if (!(path.Length > 0)) return;
            int count = c["boneIds"].Items.Count; var indices = new int[count]; var lengths = new float[count]; var snapshot = (Affine[])pose.World.Clone();
            bool tangent = c.S("rotateMode") == "tangent", chainScale = c.S("rotateMode") == "chainScale"; int spacesCount = tangent ? count : count + 1;
            float total = 0;
            for (int i = 0; i < count; i++)
            {
                int bone = pose.Data.BoneIndex[c["boneIds"][i].String]; indices[i] = bone; lengths[i] = Math.Max(0, pose.Data.BoneData[bone].Length) * Solvers.Sx(snapshot[bone]);
                if (i < (tangent ? count - 1 : count)) total += lengths[i];
            }
            float spacing = c.F("spacing"), distance = c.F("position") * (c.S("positionMode") == "percent" ? path.Length : 1);
            float rotation = c.F("rotation"), offset = rotation != 0 && snapshot[targetBone].Determinant < 0 ? -rotation : rotation;
            float mr = Numeric.Clamp(c.F("mixRotate", 1), 0, 1), mx = Numeric.Clamp(c.F("mixX", 1), 0, 1), my = Numeric.Clamp(c.F("mixY", 1), 0, 1);
            var desired = new Dictionary<int, Affine>(); PathSampler.Sample? previous = null;
            float? translationResidual = null, rotationResidual = null, scaleResidual = null;
            for (int i = 0; i < count; i++)
            {
                int bone = indices[i]; float setupLength = Math.Max(0, pose.Data.BoneData[bone].Length), increment;
                switch (c.S("spacingMode"))
                {
                    case "length": increment = setupLength > 0 ? Math.Max(setupLength + spacing, 0) * lengths[i] / setupLength : spacing; break;
                    case "fixed": increment = setupLength > 0 ? spacing * lengths[i] / setupLength : spacing; break;
                    case "percent": increment = spacing * path.Length; break;
                    default: increment = total > 0 ? (setupLength > 0 ? lengths[i] : spacing) / total * spacing * path.Length : spacing * path.Length / Math.Max(spacesCount, 1); break;
                }
                PathSampler.Sample? sample = previous ?? path.At(distance), next = path.At(distance + increment); previous = null; distance += increment;
                if (!sample.HasValue || !next.HasValue) continue;
                Point point = sample.Value.Position, ahead = next.Value.Position; Affine current = snapshot[bone]; float oldAngle = Solvers.Angle(current);
                float chainAngle = Numeric.Atan2(ahead.Y - point.Y, ahead.X - point.X) * Numeric.RadToDeg;
                float angle = oldAngle + Numeric.Wrap((tangent ? sample.Value.Tangent : chainAngle) + offset - oldAngle) * mr;
                float scaleX = Solvers.Sx(current), scaleY = Solvers.Sy(current);
                if (chainScale && setupLength != 0) scaleX += (Numeric.Hypot(ahead.X - point.X, ahead.Y - point.Y) / setupLength - scaleX) * mr;
                float shearY = Solvers.YAngle(current) - oldAngle - 90;
                Affine nextWorld = Affine.FromLocal(new BoneLocal(current.Tx + (point.X - current.Tx) * mx, current.Ty + (point.Y - current.Ty) * my, angle, scaleX, scaleY, 0, shearY));
                desired.Add(bone, nextWorld);
                if (mx != 0 || my != 0) Solvers.IncludeMax(ref translationResidual, Numeric.Hypot(mx == 0 ? 0 : nextWorld.Tx - point.X, my == 0 ? 0 : nextWorld.Ty - point.Y));
                if (mr != 0) Solvers.IncludeMax(ref rotationResidual, Math.Abs(Numeric.Wrap(Solvers.Angle(nextWorld) - (tangent ? sample.Value.Tangent : chainAngle) - offset)));
                if (chainScale && mr != 0 && setupLength != 0) Solvers.IncludeMax(ref scaleResidual, Math.Abs(Solvers.Sx(nextWorld) - Numeric.Hypot(ahead.X - point.X, ahead.Y - point.Y) / setupLength));
                if (!tangent && !chainScale && rotation == 0 && mr != 0)
                {
                    Affine rotated = Solvers.Rotate(current, Numeric.Wrap(chainAngle - oldAngle));
                    Point tip = new Point(point.X + setupLength * rotated.A, point.Y + setupLength * rotated.B);
                    previous = new PathSampler.Sample(new Point(ahead.X + (tip.X - ahead.X) * mr, ahead.Y + (tip.Y - ahead.Y) * mr), next.Value.Tangent);
                }
            }
            var affected = new bool[pose.World.Length];
            for (int i = 0; i < affected.Length; i++)
            {
                int parent = pose.Data.BoneData[i].Parent;
                if (desired.TryGetValue(i, out Affine matrix)) { affected[i] = true; Solvers.SetWorld(pose, i, matrix, root); }
                else if (parent >= 0 && affected[parent]) { affected[i] = true; pose.World[i] = Affine.Child(pose.World[parent], pose.Locals[i], pose.Modes[i]); }
            }
            pose.Diagnostics[c.S("id")] = new PathConstraintDiagnostic(desired.Count, mr, mx, my, translationResidual, rotationResidual, scaleResidual);
        }
    }
}
